src/compiler/checker.ts

```diff
@@ -4297,14 +4300,36 @@ namespace ts {
                 }
                 else if (!jsDocType) {
                     // If we don't have an explicit JSDoc type, get the type from the expression.
-                    types.push(getWidenedLiteralType(checkExpressionCached(expression.right)));
+                    const type = getWidenedLiteralType(checkExpressionCached(expression.right));
+                    let anyedType = type;
+                    if (isEmptyArrayLiteralType(type)) {
+                        anyedType = anyArrayType;
+                        if (noImplicitAny) {
+                            reportImplicitAnyError(expression, anyArrayType);
+                        }
+                    }
+                    types.push(anyedType);
+                    if (declarationInConstructor) {
+                        (constructorTypes || (constructorTypes = [])).push(anyedType);
+                    }
                 }
             }
-
-            const type = jsDocType || getUnionType(types, UnionReduction.Subtype);
-            return getWidenedType(addOptionality(type, definedInMethod && !definedInConstructor));
+            let type = jsDocType;
+            if (!type) {
+                const sourceTypes = some(constructorTypes, t => !!(t.flags & ~(TypeFlags.Nullable | TypeFlags.ContainsWideningType))) ? constructorTypes : types;
+                type = getUnionType(sourceTypes, UnionReduction.Subtype);
+            }
+            const widened = getWidenedType(addOptionality(type, definedInMethod && !definedInConstructor));
+            if (filterType(widened, t => !!(t.flags & ~TypeFlags.Nullable)) === neverType) {

```

Comment:
This isn't going to work for generics extending the "nullable" types is it? Eg

```ts
// @filename: m.d.ts
type IsNullable<T extends undefined | null> = T;
type AsNullable<T> = [T] extends [IsNullable<infer U>] ? U : never;
declare function foo<T>(x: T, y: (p: T) => AsNullable<T>): AsNullable<T>;
// @filename: file.js
/**
* @template {T}
* @param {T} x
*/
function Thing(x) {
  this.member = foo(x, x => undefined);
}
```
? (Without generic bounds in js this was the only way I could think to introduce a bounded generic inside js, so maybe this is less of an issue and more of a statement?)